// @n8n/n8n-nodes-langchain/dist/nodes/text_splitters/TextSplitterRecursiveCharacterTextSplitter/TextSplitterRecursiveCharacterTextSplitter.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextSplitterRecursiveCharacterTextSplitter = void 0;
const text_splitter_1 = require("langchain/text_splitter");
const logWrapper_1 = require("../../../utils/logWrapper");
class TextSplitterRecursiveCharacterTextSplitter {
    constructor() {
        this.description = {
            displayName: 'Recursive Character Text Splitter',
            name: 'textSplitterRecursiveCharacterTextSplitter',
            icon: 'fa:grip-lines-vertical',
            group: ['transform'],
            version: 1,
            description: 'Split text into chunks by characters recursively, recommended for most use cases',
            defaults: {
                name: 'Recursive Character Text Splitter',
            },
            codex: {
                categories: ['AI'],
                subcategories: {
                    AI: ['Text Splitters'],
                },
                resources: {
                    primaryDocumentation: [
                        {
                            url: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.textsplitterrecursivecharactertextsplitter/',
                        },
                    ],
                },
            },
            inputs: [],
            outputs: ["ai_textSplitter"],
            outputNames: ['Text Splitter'],
            properties: [
                {
                    displayName: 'Chunk Size',
                    name: 'chunkSize',
                    type: 'number',
                    default: 1000,
                },
                {
                    displayName: 'Chunk Overlap',
                    name: 'chunkOverlap',
                    type: 'number',
                    default: 0,
                },
            ],
        };
    }
    async supplyData(itemIndex) {
        this.logger.verbose('Supply Data for Text Splitter');
        const chunkSize = this.getNodeParameter('chunkSize', itemIndex);
        const chunkOverlap = this.getNodeParameter('chunkOverlap', itemIndex);
        const params = {
            separators: ['\n\n', '\n', ' ', ''],
            chunkSize,
            chunkOverlap,
            keepSeparator: false,
        };
        const splitter = new text_splitter_1.RecursiveCharacterTextSplitter(params);
        return {
            response: (0, logWrapper_1.logWrapper)(splitter, this),
        };
    }
}
exports.TextSplitterRecursiveCharacterTextSplitter = TextSplitterRecursiveCharacterTextSplitter;
