// @n8n/n8n-nodes-langchain/dist/credentials/CohereApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CohereApi = void 0;
class CohereApi {
    constructor() {
        this.name = 'cohereApi';
        this.displayName = 'CohereApi';
        this.documentationUrl = 'cohere';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://api.cohere.ai',
                url: '/v1/detect-language',
                method: 'POST',
                body: {
                    texts: ['hello'],
                },
            },
        };
    }
}
exports.CohereApi = CohereApi;
