// @n8n/n8n-nodes-langchain/dist/credentials/PineconeApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PineconeApi = void 0;
class PineconeApi {
    constructor() {
        this.name = 'pineconeApi';
        this.displayName = 'PineconeApi';
        this.documentationUrl = 'pinecone';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
            {
                displayName: 'Environment',
                name: 'environment',
                type: 'string',
                required: true,
                default: 'us-central1-gcp',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'Api-Key': '={{$credentials.apiKey}}',
                },
            },
        };
    }
}
exports.PineconeApi = PineconeApi;
