// @n8n/n8n-nodes-langchain/dist/nodes/memory/MemoryZep/MemoryZep.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MemoryZep = void 0;
const zep_1 = require("langchain/memory/zep");
const logWrapper_1 = require("../../../utils/logWrapper");
class MemoryZep {
    constructor() {
        this.description = {
            displayName: 'Zep',
            name: 'memoryZep',
            icon: 'file:zep.png',
            group: ['transform'],
            version: 1,
            description: 'Use Zep Memory',
            defaults: {
                name: 'Zep',
            },
            codex: {
                categories: ['AI'],
                subcategories: {
                    AI: ['Memory'],
                },
                resources: {
                    primaryDocumentation: [
                        {
                            url: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.memoryzep/',
                        },
                    ],
                },
            },
            inputs: [],
            outputs: ["ai_memory"],
            outputNames: ['Memory'],
            credentials: [
                {
                    name: 'zepApi',
                    required: true,
                },
            ],
            properties: [
                {
                    displayName: 'Session ID',
                    name: 'sessionId',
                    type: 'string',
                    required: true,
                    default: '',
                },
            ],
        };
    }
    async supplyData(itemIndex) {
        const credentials = (await this.getCredentials('zepApi'));
        const sessionId = this.getNodeParameter('sessionId', itemIndex);
        const memory = new zep_1.ZepMemory({
            sessionId,
            baseURL: credentials.apiUrl,
            apiKey: credentials.apiKey,
            memoryKey: 'chat_history',
            returnMessages: true,
            inputKey: 'input',
            outputKey: 'output',
        });
        return {
            response: (0, logWrapper_1.logWrapper)(memory, this),
        };
    }
}
exports.MemoryZep = MemoryZep;
