// @n8n/n8n-nodes-langchain/dist/nodes/tools/ToolWorkflow/ToolWorkflow.node.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolWorkflow = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const manual = __importStar(require("n8n-nodes-base/dist/nodes/Set/v2/manual.mode"));
const tools_1 = require("langchain/tools");
const get_1 = __importDefault(require("lodash/get"));
const isObject_1 = __importDefault(require("lodash/isObject"));
class ToolWorkflow {
    constructor() {
        this.description = {
            displayName: 'Workflow Tool',
            name: 'toolWorkflow',
            icon: 'fa:network-wired',
            group: ['transform'],
            version: 1,
            description: 'Create a tool via a workflow',
            defaults: {
                name: 'Workflow Tool',
            },
            codex: {
                categories: ['AI'],
                subcategories: {
                    AI: ['Tools'],
                },
                resources: {
                    primaryDocumentation: [
                        {
                            url: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.toolworkflow/',
                        },
                    ],
                },
            },
            inputs: [],
            outputs: ["ai_tool"],
            outputNames: ['Tool'],
            properties: [
                {
                    displayName: 'Name',
                    name: 'name',
                    type: 'string',
                    default: '',
                    placeholder: 'My_Color_Tool',
                },
                {
                    displayName: 'Description',
                    name: 'description',
                    type: 'string',
                    default: '',
                    placeholder: 'Call this tool to get a random color. The input should be a string with comma separted names of colors to exclude.',
                    typeOptions: {
                        rows: 3,
                    },
                },
                {
                    displayName: 'The workflow will receive "query" as input and the output of the last node will be returned as response',
                    name: 'executeNotice',
                    type: 'notice',
                    default: '',
                },
                {
                    displayName: 'Source',
                    name: 'source',
                    type: 'options',
                    options: [
                        {
                            name: 'Database',
                            value: 'database',
                            description: 'Load the workflow from the database by ID',
                        },
                        {
                            name: 'Parameter',
                            value: 'parameter',
                            description: 'Load the workflow from a parameter',
                        },
                    ],
                    default: 'database',
                    description: 'Where to get the workflow to execute from',
                },
                {
                    displayName: 'Workflow ID',
                    name: 'workflowId',
                    type: 'string',
                    displayOptions: {
                        show: {
                            source: ['database'],
                        },
                    },
                    default: '',
                    required: true,
                    description: 'The workflow to execute',
                },
                {
                    displayName: 'Workflow JSON',
                    name: 'workflowJson',
                    type: 'string',
                    typeOptions: {
                        editor: 'json',
                        rows: 10,
                    },
                    displayOptions: {
                        show: {
                            source: ['parameter'],
                        },
                    },
                    default: '\n\n\n',
                    required: true,
                    description: 'The workflow JSON code to execute',
                },
                {
                    displayName: 'Response Property Name',
                    name: 'responsePropertyName',
                    type: 'string',
                    default: 'response',
                    description: 'The name of the property of the last node that will be returned as response',
                },
                {
                    displayName: 'Workflow Values',
                    name: 'fields',
                    placeholder: 'Add Value',
                    type: 'fixedCollection',
                    description: 'Set the values which should be made available in the workflow',
                    typeOptions: {
                        multipleValues: true,
                        sortable: true,
                    },
                    default: {},
                    options: [
                        {
                            name: 'values',
                            displayName: 'Values',
                            values: [
                                {
                                    displayName: 'Name',
                                    name: 'name',
                                    type: 'string',
                                    default: '',
                                    placeholder: 'e.g. fieldName',
                                    description: 'Name of the field to set the value of. Supports dot-notation. Example: data.person[0].name.',
                                    requiresDataPath: 'single',
                                },
                                {
                                    displayName: 'Type',
                                    name: 'type',
                                    type: 'options',
                                    description: 'The field value type',
                                    options: [
                                        {
                                            name: 'String',
                                            value: 'stringValue',
                                        },
                                        {
                                            name: 'Number',
                                            value: 'numberValue',
                                        },
                                        {
                                            name: 'Boolean',
                                            value: 'booleanValue',
                                        },
                                        {
                                            name: 'Array',
                                            value: 'arrayValue',
                                        },
                                        {
                                            name: 'Object',
                                            value: 'objectValue',
                                        },
                                    ],
                                    default: 'stringValue',
                                },
                                {
                                    displayName: 'Value',
                                    name: 'stringValue',
                                    type: 'string',
                                    default: '',
                                    displayOptions: {
                                        show: {
                                            type: ['stringValue'],
                                        },
                                    },
                                    validateType: 'string',
                                    ignoreValidationDuringExecution: true,
                                },
                                {
                                    displayName: 'Value',
                                    name: 'numberValue',
                                    type: 'string',
                                    default: '',
                                    displayOptions: {
                                        show: {
                                            type: ['numberValue'],
                                        },
                                    },
                                    validateType: 'number',
                                    ignoreValidationDuringExecution: true,
                                },
                                {
                                    displayName: 'Value',
                                    name: 'booleanValue',
                                    type: 'options',
                                    default: 'true',
                                    options: [
                                        {
                                            name: 'True',
                                            value: 'true',
                                        },
                                        {
                                            name: 'False',
                                            value: 'false',
                                        },
                                    ],
                                    displayOptions: {
                                        show: {
                                            type: ['booleanValue'],
                                        },
                                    },
                                    validateType: 'boolean',
                                    ignoreValidationDuringExecution: true,
                                },
                                {
                                    displayName: 'Value',
                                    name: 'arrayValue',
                                    type: 'string',
                                    default: '',
                                    placeholder: 'e.g. [ arrayItem1, arrayItem2, arrayItem3 ]',
                                    displayOptions: {
                                        show: {
                                            type: ['arrayValue'],
                                        },
                                    },
                                    validateType: 'array',
                                    ignoreValidationDuringExecution: true,
                                },
                                {
                                    displayName: 'Value',
                                    name: 'objectValue',
                                    type: 'string',
                                    default: '={}',
                                    typeOptions: {
                                        editor: 'json',
                                        editorLanguage: 'json',
                                        rows: 2,
                                    },
                                    displayOptions: {
                                        show: {
                                            type: ['objectValue'],
                                        },
                                    },
                                    validateType: 'object',
                                    ignoreValidationDuringExecution: true,
                                },
                            ],
                        },
                    ],
                },
            ],
        };
    }
    async supplyData(itemIndex) {
        const name = this.getNodeParameter('name', itemIndex);
        const description = this.getNodeParameter('description', itemIndex);
        const runFunction = async (query) => {
            const source = this.getNodeParameter('source', itemIndex);
            const responsePropertyName = this.getNodeParameter('responsePropertyName', itemIndex);
            const workflowInfo = {};
            if (source === 'database') {
                workflowInfo.id = this.getNodeParameter('workflowId', itemIndex);
            }
            else if (source === 'parameter') {
                const workflowJson = this.getNodeParameter('workflowJson', itemIndex);
                try {
                    workflowInfo.code = JSON.parse(workflowJson);
                }
                catch (error) {
                    throw new n8n_workflow_1.NodeOperationError(this.getNode(), `The provided workflow is not valid JSON: "${error.message}"`, {
                        itemIndex,
                    });
                }
            }
            const rawData = { query };
            const workflowFieldsJson = this.getNodeParameter('fields.values', itemIndex, [], {
                rawExpressions: true,
            });
            for (const entry of workflowFieldsJson) {
                if (entry.type === 'objectValue' && entry.objectValue.startsWith('=')) {
                    rawData[entry.name] = entry.objectValue.replace(/^=+/, '');
                }
            }
            const options = {
                include: 'all',
            };
            const newItem = await manual.execute.call(this, { json: { query } }, itemIndex, options, rawData, this.getNode());
            const items = [newItem];
            let receivedData;
            try {
                receivedData = (await this.executeWorkflow(workflowInfo, items));
            }
            catch (error) {
                throw new n8n_workflow_1.NodeOperationError(this.getNode(), error);
            }
            const response = (0, get_1.default)(receivedData, [
                0,
                0,
                'json',
                responsePropertyName,
            ]);
            if (response === undefined) {
                throw new n8n_workflow_1.NodeOperationError(this.getNode(), `There was an error: "The workflow did not return an item with the property '${responsePropertyName}'"`);
            }
            return response;
        };
        return {
            response: new tools_1.DynamicTool({
                name,
                description,
                func: async (query) => {
                    const { index } = this.addInputData("ai_tool", [[{ json: { query } }]]);
                    let response = '';
                    let executionError;
                    try {
                        response = await runFunction(query);
                    }
                    catch (error) {
                        executionError = error;
                        response = `There was an error: "${error.message}"`;
                    }
                    if (typeof response === 'number') {
                        response = response.toString();
                    }
                    if ((0, isObject_1.default)(response)) {
                        response = JSON.stringify(response, null, 2);
                    }
                    if (typeof response !== 'string') {
                        executionError = new n8n_workflow_1.NodeOperationError(this.getNode(), `The code did not return a valid value. Instead of a string did a value of type '${typeof response}' get returned.`);
                        response = `There was an error: "${executionError.message}"`;
                    }
                    if (executionError) {
                        void this.addOutputData("ai_tool", index, executionError);
                    }
                    else {
                        void this.addOutputData("ai_tool", index, [[{ json: { response } }]]);
                    }
                    return response;
                },
            }),
        };
    }
}
exports.ToolWorkflow = ToolWorkflow;
