// @n8n/n8n-nodes-langchain/dist/nodes/tools/ToolWikipedia/ToolWikipedia.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolWikipedia = void 0;
const tools_1 = require("langchain/tools");
const logWrapper_1 = require("../../../utils/logWrapper");
class ToolWikipedia {
    constructor() {
        this.description = {
            displayName: 'Wikipedia',
            name: 'toolWikipedia',
            icon: 'file:wikipedia.svg',
            group: ['transform'],
            version: 1,
            description: 'Search in Wikipedia',
            defaults: {
                name: 'Wikipedia',
            },
            codex: {
                categories: ['AI'],
                subcategories: {
                    AI: ['Tools'],
                },
                resources: {
                    primaryDocumentation: [
                        {
                            url: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.toolwikipedia/',
                        },
                    ],
                },
            },
            inputs: [],
            outputs: ["ai_tool"],
            outputNames: ['Tool'],
            properties: [],
        };
    }
    async supplyData() {
        return {
            response: (0, logWrapper_1.logWrapper)(new tools_1.WikipediaQueryRun(), this),
        };
    }
}
exports.ToolWikipedia = ToolWikipedia;
