// @n8n/n8n-nodes-langchain/dist/nodes/llms/LMOpenHuggingFaceInference/HuggingFaceInferenceEndpoints.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HuggingFaceInference = void 0;
const base_1 = require("langchain/llms/base");
class HuggingFaceInference extends base_1.LLM {
    get lc_secrets() {
        return {
            apiKey: 'HUGGINGFACEHUB_API_KEY',
        };
    }
    constructor(fields) {
        var _a, _b, _c, _d, _e, _f;
        super(fields !== null && fields !== void 0 ? fields : {});
        this.model = 'gpt2';
        this.temperature = undefined;
        this.maxTokens = undefined;
        this.topP = undefined;
        this.topK = undefined;
        this.frequencyPenalty = undefined;
        this.apiKey = undefined;
        this.endpointUrl = undefined;
        this.model = (_a = fields === null || fields === void 0 ? void 0 : fields.model) !== null && _a !== void 0 ? _a : this.model;
        this.temperature = (_b = fields === null || fields === void 0 ? void 0 : fields.temperature) !== null && _b !== void 0 ? _b : this.temperature;
        this.maxTokens = (_c = fields === null || fields === void 0 ? void 0 : fields.maxTokens) !== null && _c !== void 0 ? _c : this.maxTokens;
        this.topP = (_d = fields === null || fields === void 0 ? void 0 : fields.topP) !== null && _d !== void 0 ? _d : this.topP;
        this.topK = (_e = fields === null || fields === void 0 ? void 0 : fields.topK) !== null && _e !== void 0 ? _e : this.topK;
        this.frequencyPenalty = (_f = fields === null || fields === void 0 ? void 0 : fields.frequencyPenalty) !== null && _f !== void 0 ? _f : this.frequencyPenalty;
        this.apiKey = fields === null || fields === void 0 ? void 0 : fields.apiKey;
        this.endpointUrl = fields === null || fields === void 0 ? void 0 : fields.endpointUrl;
        if (!this.apiKey) {
            throw new Error('Please set an API key for HuggingFace Hub in the environment variable HUGGINGFACEHUB_API_KEY or in the apiKey field of the HuggingFaceInference constructor.');
        }
    }
    _llmType() {
        return 'hf';
    }
    async _call(prompt, options) {
        const { HfInference } = await HuggingFaceInference.imports();
        const hf = this.endpointUrl
            ? new HfInference(this.apiKey).endpoint(this.endpointUrl)
            : new HfInference(this.apiKey);
        const res = await this.caller.callWithOptions({ signal: options.signal }, hf.textGeneration.bind(hf), {
            model: this.model,
            parameters: {
                return_full_text: false,
                temperature: this.temperature,
                max_new_tokens: this.maxTokens,
                top_p: this.topP,
                top_k: this.topK,
                repetition_penalty: this.frequencyPenalty,
            },
            inputs: prompt,
        });
        return res.generated_text;
    }
    static async imports() {
        try {
            const { HfInference } = await Promise.resolve().then(() => __importStar(require('@huggingface/inference')));
            return { HfInference };
        }
        catch (e) {
            throw new Error('Please install huggingface as a dependency with, e.g. `yarn add @huggingface/inference`');
        }
    }
}
exports.HuggingFaceInference = HuggingFaceInference;
