// @n8n/n8n-nodes-langchain/dist/nodes/memory/MemoryMotorhead/MemoryMotorhead.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MemoryMotorhead = void 0;
const memory_1 = require("langchain/memory");
const logWrapper_1 = require("../../../utils/logWrapper");
class MemoryMotorhead {
    constructor() {
        this.description = {
            displayName: 'Motorhead',
            name: 'memoryMotorhead',
            icon: 'fa:file-export',
            group: ['transform'],
            version: 1,
            description: 'Use Motorhead Memory',
            defaults: {
                name: 'Motorhead',
            },
            codex: {
                categories: ['AI'],
                subcategories: {
                    AI: ['Memory'],
                },
                resources: {
                    primaryDocumentation: [
                        {
                            url: 'https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.memorymotorhead/',
                        },
                    ],
                },
            },
            inputs: [],
            outputs: ["ai_memory"],
            outputNames: ['Memory'],
            credentials: [
                {
                    name: 'motorheadApi',
                    required: true,
                },
            ],
            properties: [
                {
                    displayName: 'Session ID',
                    name: 'sessionId',
                    type: 'string',
                    required: true,
                    default: '',
                },
            ],
        };
    }
    async supplyData(itemIndex) {
        const credentials = await this.getCredentials('motorheadApi');
        const sessionId = this.getNodeParameter('sessionId', itemIndex);
        const memory = new memory_1.MotorheadMemory({
            sessionId,
            url: `${credentials.host}/motorhead`,
            clientId: credentials.clientId,
            apiKey: credentials.apiKey,
            memoryKey: 'chat_history',
            returnMessages: true,
        });
        await memory.init();
        return {
            response: (0, logWrapper_1.logWrapper)(memory, this),
        };
    }
}
exports.MemoryMotorhead = MemoryMotorhead;
